src/ca/verify.ts

```diff
@@ -0,0 +1,125 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+import { VerificationError } from '../error';
+import * as sigstore from '../types/sigstore';
+import { x509Certificate } from '../x509/cert';
+import { verifyCertificateChain } from '../x509/verify';
+
+export function verifySigningCertificate(
+  bundle: sigstore.Bundle,
+  trustedRoot: sigstore.TrustedRoot,
+  options: sigstore.ArtifactVerificationOptions_CtlogOptions
+) {
+  if (bundle.verificationMaterial?.content?.$case !== 'x509CertificateChain') {
+    throw new VerificationError('No certificate chain in bundle');
+  }
+
+  const bundleCerts = parseCerts(
+    bundle.verificationMaterial.content.x509CertificateChain.certificates
+  );
+
+  // Check that a trusted certificate chain can be found for the signing
+  // certificate in the bundle
+  const trustedChain = verifyChain(
+    bundleCerts,
+    trustedRoot.certificateAuthorities
+  );
+
+  // Unless disabled, verify the SCTs in the signing certificate
+  if (!options.disable) {
+    verifySCTs(trustedChain, trustedRoot.ctlogs, options);
+  }
+}
+
+function verifyChain(
+  bundleCerts: x509Certificate[],
+  certificateAuthorities: sigstore.CertificateAuthority[]
+): x509Certificate[] {
+  const signingCert = bundleCerts[0];
+
+  // Filter the list of certificate authorities to those which are valid for the
+  // signing certificate's notBefore date.
+  const validCAs = filterCertificateAuthorities(
+    certificateAuthorities,
+    signingCert.notBefore
+  );
+
+  if (validCAs.length === 0) {
+    throw new VerificationError('No valid certificate authorities');
+  }
+
+  let trustedChain: x509Certificate[] = [];
+
+  // Loop through all valid CAs and attempt to verify the certificate chain
+  const verified = validCAs.find((ca) => {
+    const trustedCerts = parseCerts(ca.certChain?.certificates || []);
+    try {
+      trustedChain = verifyCertificateChain({
+        trustedCerts,
+        certs: bundleCerts,
+        validAt: signingCert.notBefore,
+      });
+      return true;
+    } catch (e) {
+      return false;
+    }
+  });
+
+  if (!verified) {
+    throw new VerificationError('No valid certificate chain');
+  }
+
+  return trustedChain;
+}
+
+function verifySCTs(
+  certificateChain: x509Certificate[],
+  ctLogs: sigstore.TransparencyLogInstance[],
+  options: sigstore.ArtifactVerificationOptions_CtlogOptions
+) {
+  const signingCert = certificateChain[0];
+  const issuerCert = certificateChain[1];
+  const sctResults = signingCert.verifySCTs(issuerCert, ctLogs);
+
+  // Count the number of verified SCTs which were found
+  const verifiedSCTCount = sctResults.filter((sct) => sct.verified).length;
+
+  if (verifiedSCTCount < options.threshold) {
```

Comment:
Maybe ok, but wondering if this should be more explicit like `SCTVerificationThreshold `?

Reply:
Ref https://github.com/sigstore/sigstore-js/pull/248#discussion_r1066013245

Reply:
Agree that the naming is vague, but this is another case where we're using the types generated directly from the protobufs (https://github.com/sigstore/protobuf-specs/blob/a09e324bbc409628cef99a33e05fad2a5b4c47e5/protos/sigstore_verification.proto#L57-L66)